### ISSUE TITLE
[proposal] Dockerfile `ENV --from=builder BAR=$FOO` to access previous stage ENV vars

### ISSUE THREAD

**Issue by mikecook** [moby/moby#36941]
TLDR: Allow ENV to use `--from=builder` to access previous stage ENV vars in the same way that COPY uses `--from=builder` to access previous stage files. `ENV --from=builder VAR=$VAR_FROM_BUILD_STAGE`

Multistage builds break setting non-static environment variables within the Dockerfile, since ENV vars are not passed from the build stage to any later stages without ugly workarounds like writing files, copying the files, and then injecting them into the CMD/ENTRYPOINT.

I understand that docker-compose has an .env file solution, and that `docker run` allows you to set environment variables. What I would like is `docker build` allowing me to also set ENV variables either from the command line or the builder stage.

In a single stage build, ENV vars stay set:
```
$ cat Dockerfile.singlestage
FROM alpine
ARG FOO
ENV BAR=$FOO
$ docker build -t test -f Dockerfile.singlestage --build-arg FOO=TEST .
Sending build context to Docker daemon  182.6MB
Step 1/3 : FROM alpine
 ---> 3fd9065eaf02
Step 2/3 : ARG FOO
 ---> Using cache
 ---> 0d113aa1d5e0
Step 3/3 : ENV BAR=$FOO
 ---> Using cache
 ---> 5a11d889e2f0
Successfully built 5a11d889e2f0
Successfully tagged test:latest
$ docker run --rm -it test env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=05544e76df3f
TERM=xterm
BAR=TEST    # <----------------------------- Our test var
HOME=/root
```

In a multistage build, ENV vars set during build stages never make it to the final stage
```
$ cat Dockerfile.multistage
FROM alpine as builder
ARG FOO
ENV BAR=$FOO

FROM alpine
ENV BAZ=$BAR
$ docker build -t test -f Dockerfile.multistage --build-arg FOO=TEST .
Sending build context to Docker daemon  182.6MB
Step 1/5 : FROM alpine as builder
 ---> 3fd9065eaf02
Step 2/5 : ARG FOO
 ---> Using cache
 ---> 0d113aa1d5e0
Step 3/5 : ENV BAR=$FOO
 ---> Using cache
 ---> 5a11d889e2f0
Step 4/5 : FROM alpine
 ---> 3fd9065eaf02
Step 5/5 : ENV BAZ=$BAR
 ---> Using cache
 ---> 08363f85680c
Successfully built 08363f85680c
Successfully tagged test:latest
$ docker run --rm -it test env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=8535c760799c
TERM=xterm
BAZ=        # <----------------------------- Our test var
HOME=/root
```

How I would like it to work. Variable populated via request from previous stage image:
```
$ cat Dockerfile.wish
FROM alpine as builder
ARG FOO
ENV BAR=$FOO

FROM alpine
ENV --from=builder BAZ=$BAR
$ docker build -t test -f Dockerfile.wish --build-arg FOO=TEST .
Sending build context to Docker daemon  182.6MB
Step 1/5 : FROM alpine as builder
 ---> 3fd9065eaf02
Step 2/5 : ARG FOO
 ---> Using cache
 ---> 0d113aa1d5e0
Step 3/5 : ENV BAR=$FOO
 ---> Using cache
 ---> 5a11d889e2f0
Step 4/5 : FROM alpine
 ---> 3fd9065eaf02
Step 5/5 : ENV --from=builder BAZ=$BAR
 ---> Using cache
 ---> 08363f85680c
Successfully built 08363f85680c
Successfully tagged test:latest
$ docker run --rm -it test env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=8535c760799c
TERM=xterm
BAZ=TEST     # <----------------------------- Our test var
HOME=/root

```

**Comment by belldoor** [comment]
+1
